File: src/demo_python_tf/demo_python_tf/tf_listener.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformListener, Buffer
from tf_transformations import euler_from_quaternion
import math


class TFListener(Node):
    def __init__(self):
        super().__init__('tf_listener')
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.timer = self.create_timer(1.0, self.lookup_transform)  # Check for transform at 1 Hz

    def lookup_transform(self):
        try:
            t = self.tf_buffer.lookup_transform('base_link', 'bottle_link', rclpy.time.Time(seconds=0.0), rclpy.time.Duration(seconds=1.0))
            transform=t.transform
            translation = transform.translation
            self.get_logger().info(f'平移 from base_link to bottle_link: {translation}')
            rotation = transform.rotation
            self.get_logger().info(f'旋转 from base_link to bottle_link: {rotation}')
            euler = euler_from_quaternion([rotation.x, rotation.y, rotation.z, rotation.w])
            self.get_logger().info(f'欧拉角 from base_link to bottle_link: roll={math.degrees(euler[0])}, pitch={math.degrees(euler[1])}, yaw={math.degrees(euler[2])}')
            self.get_logger().info(f'Got transform from camera_link to bottle_link: translation={translation}, rotation={euler}')
        except Exception as e:
            self.get_logger().warn(f'Could not get transform: {e}')


def main(args=None):
    rclpy.init(args=args)
    node = TFListener()
    rclpy.spin(node)
    rclpy.shutdown()

File: src/demo_cpp_tf/src/tf_listener.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/utils.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "chrono"

using namespace std::chrono_literals;

class TFListener : public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
    rclcpp::TimerBase::SharedPtr timer_;
public:
    TFListener() : Node("tf_listener")
    {
        tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_, this);
        timer_ = this->create_wall_timer(1s, std::bind(&TFListener::timer_callback, this));
    }

    void timer_callback()
    {
        geometry_msgs::msg::TransformStamped transformStamped;
        try {
            transformStamped = tf_buffer_->lookupTransform("base_link", "target_point", this->get_clock()->now(), rclcpp::Duration::from_seconds(1.0f));
            RCLCPP_INFO(this->get_logger(), "Transform from map to base_link: translation(%.2f, %.2f, %.2f), rotation(%.2f, %.2f, %.2f, %.2f)",
                transformStamped.transform.translation.x,
                transformStamped.transform.translation.y,
                transformStamped.transform.translation.z,
                transformStamped.transform.rotation.x,
                transformStamped.transform.rotation.y,
                transformStamped.transform.rotation.z,
                transformStamped.transform.rotation.w);
            auto translation = transformStamped.transform.translation;
            auto rotation = transformStamped.transform.rotation;
            double roll, pitch, yaw;
            tf2::getEulerYPR(rotation, yaw, pitch, roll);
            RCLCPP_INFO(this->get_logger(), "平移 Translation: x=%.2f, y=%.2f, z=%.2f", translation.x, translation.y, translation.z);
            RCLCPP_INFO(this->get_logger(), "旋转 Euler angles: roll=%.2f, pitch=%.2f, yaw=%.2f", roll, pitch, yaw);
        } catch (tf2::TransformException & ex) {
            RCLCPP_WARN(this->get_logger(), "Could not get transform: %s", ex.what());
        }
    }
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TFListener>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}